### PR TITLE
add function input and output parameters...

### PR DESCRIPTION
**Pull request by fkromer**
to "Programming in D for C++ Programmers"

### CODE REVIEW

File: cpptod.dd [hunk 1]
```diff
@@ -829,6 +831,106 @@ void test()
 }
 ------
 
+<hr>$(COMMENT  -------------------------------------------- )
+
+$(H3 <a name="functioninputparameters">Function Input Parameters</a>)
```
Comment (wilzbach) [inline cpptod.dd]: Why not `$(LNAME2 functioninputparameters, Function Input Parameters)` (the DDoc macros are a hard to getting used too).

File: cpptod.dd [hunk 2]
```diff
@@ -829,6 +831,106 @@ void test()
...
+function(int& outputArgument);
+// outputArgument = 42
+)
+
+$(H4 The D Way)
+
+	D supports explicit output parameters.
+
+------
+void function(out int outputParameter) {
```
Comment (wilzbach) [inline cpptod.dd]: Maybe `foo` as a name to avoid confusing below?

File: cpptod.dd [hunk 3]
```diff
@@ -829,6 +831,106 @@ void test()
...
+    outputParameter = 42;
+}
+------
+
+The function is used as follows:
+
+------
+int outputArgument = 0;
+function(outputArgument);
+// outputArgument = 42
```
Comment (wilzbach) [inline cpptod.dd]: `assert(outputArgument == 42);`

If you want, you can do:

```
$(RUNNABLE_EXAMPLE
---
void main()
  // <- your code here
}
---
)
```

This has the advantage that people can play with the code online.

File: cpptod.dd [hunk 4]
```diff
@@ -829,6 +831,106 @@ void test()
...
+void function(out int outputParameter) {
+    outputParameter = 42;
+}
+------
+
+The function is used as follows:
+
+------
+int outputArgument = 0;
+function(outputArgument);
```
Comment (wilzbach) [inline cpptod.dd]: Imho that's potentially confusing, hence I would use a exemplary function name above.

File: cpptod.dd [hunk 5]
```diff
@@ -829,6 +831,106 @@ void test()
 }
 ------
 
+<hr>$(COMMENT  -------------------------------------------- )
+
+$(H3 <a name="functioninputparameters">Function Input Parameters</a>)
+
+$(H4 The C++ Way)
+
+	$(D C++ does not support explicit input parameters.) If you use a
```
Comment (wilzbach) [inline cpptod.dd]: `$(D ..)` marks code D inline code:

image

Did you mean to use `$(B ...)`?

File: cpptod.dd [hunk 6]
```diff
@@ -829,6 +831,106 @@ void test()
...
+	"pass-by-value" parameter `void function(int inputOnlyImplicitly)` it can just
+	be an input parameter implicitly. If you use a "pass-by-reference" parameter
+	usually `const` is applied to the parameter
+	`void function(const int& inputOnlyParameterImplicitly)` that a forbidden
+	modification of the parameter in the function leads to a compilation error.
+
+$(CPPCODE
+void function(int implicitInputParameterPassedByValue) {
```
Comment (wilzbach) [inline cpptod.dd]: The rest of the document uses a default function name, e.g. `foo` or anything you like.

File: cpptod.dd [hunk 7]
```diff
@@ -829,6 +831,106 @@ void test()
...
+		// some modification dependent on implicitInputParameterPassedByValue
+		...
+}
+
+void function(int& implicitInputParameterPassedByReference) {
```
Comment (wilzbach) [inline cpptod.dd]: ditto

File: cpptod.dd [hunk 8]
```diff
@@ -829,6 +831,106 @@ void test()
...
+		// some modification dependent on implicitInputParameterPassedByReference
+		...
+}
+)
+
+The functions are used as follows:
+
+$(CPPCODE
+int inputArgumentPassedByValue = 42;
+function(inputArgumentPassedByValue);
```
Comment (wilzbach) [inline cpptod.dd]: ditto

File: cpptod.dd [hunk 9]
```diff
@@ -829,6 +831,106 @@ void test()
...
+// inputArgumentPassedByValue = 42
+
+int inputArgumentPassedByReference = 42;
+function(const int& inputArgumentPassedByReference);
```
Comment (wilzbach) [inline cpptod.dd]: ditto

File: cpptod.dd [hunk 10]
```diff
@@ -829,6 +831,106 @@ void test()
...
+// inputArgumentPassedByReference = 42
+)
+
+$(H4 The D Way)
+
+	D supports explicit input parameters.
+
+------
+void function(in int inputParameter) {
```
Comment (wilzbach) [inline cpptod.dd]: ditto

File: cpptod.dd [hunk 11]
```diff
@@ -829,6 +831,106 @@ void test()
...
+		// some modification dependent on inputParameter
+		...
+}
+------
+
+The function is used as follows:
+
+------
+int inputArgument = 42;
+function(inputArgument);
```
Comment (wilzbach) [inline cpptod.dd]: ditto

File: cpptod.dd [hunk 12]
```diff
@@ -829,6 +831,106 @@ void test()
...
+
+	$(D C++ does not support explicit output parameters.) You can use a reference
+	parameter to allow the function's parameter serve as input and output
+	to/from the function. (Input only parameters "called by reference" not "called
+	by value" are usually declared as such by applying `const` to the parameter
+	`void func(const int& inputParameter)`. Then the remaining uncertainty is
+	wether the parameter is an output only or an input&output paramater.)
+
+$(CPPCODE
+void function(int& inputAndOutputParameter) {
```
Comment (wilzbach) [inline cpptod.dd]: ditto

File: cpptod.dd [hunk 13]
```diff
@@ -829,6 +831,106 @@ void test()
...
+  	inputAndOutputParameter = 42;
+}
+)
+
+The function is used as follows ("call by reference"):
+
+$(CPPCODE
+int outputArgument = 0;
+function(int& outputArgument);
```
Comment (wilzbach) [inline cpptod.dd]: ditto

File: cpptod.dd [hunk 14]
```diff
@@ -829,6 +831,106 @@ void test()
...
 ------
 
+<hr>$(COMMENT  -------------------------------------------- )
+
+$(H3 <a name="functioninputparameters">Function Input Parameters</a>)
+
+$(H4 The C++ Way)
+
+	$(D C++ does not support explicit input parameters.) If you use a
+	"pass-by-value" parameter `void function(int inputOnlyImplicitly)` it can just
```
Comment (wilzbach) [inline cpptod.dd]: s/just/d

File: cpptod.dd [hunk 15]
```diff
@@ -829,6 +831,106 @@ void test()
...
+	be an input parameter implicitly. If you use a "pass-by-reference" parameter
```
Comment (wilzbach) [inline cpptod.dd]: Don't quote pass-by-reference

File: cpptod.dd [hunk 16]
```diff
@@ -829,6 +831,106 @@ void test()
...
+
+<hr>$(COMMENT  -------------------------------------------- )
+
+$(H3 <a name="functionoutputparameters">Function Output Parameters</a>)
+
+$(H4 The C++ Way)
+
+	$(D C++ does not support explicit output parameters.) You can use a reference
+	parameter to allow the function's parameter serve as input and output
+	to/from the function. (Input only parameters "called by reference" not "called
```
Comment (wilzbach) [inline cpptod.dd]: - `. (` isn't typographically allowed. Parenthesis can't be a sentence on its own
- I wouldn't quote "called by ...", maybe italic?

File: cpptod.dd [hunk 17]
```diff
@@ -829,6 +831,106 @@ void test()
...
+// inputArgument = 42
+------
+
+<hr>$(COMMENT  -------------------------------------------- )
+
+$(H3 <a name="functionoutputparameters">Function Output Parameters</a>)
+
+$(H4 The C++ Way)
+
+	$(D C++ does not support explicit output parameters.) You can use a reference
```
Comment (wilzbach) [inline cpptod.dd]: Ditto about `$(D)`

File: cpptod.dd [hunk 18]
```diff
@@ -829,6 +831,106 @@ void test()
...
+
+------
+int inputArgument = 42;
+function(inputArgument);
+// inputArgument = 42
+------
+
+<hr>$(COMMENT  -------------------------------------------- )
+
+$(H3 <a name="functionoutputparameters">Function Output Parameters</a>)
```
Comment (wilzbach) [inline cpptod.dd]: Ditto about `LNAME2`

File: cpptod.dd [hunk 19]
```diff
@@ -829,6 +831,106 @@ void test()
...
+
+$(H4 The C++ Way)
+
+	$(D C++ does not support explicit output parameters.) You can use a reference
+	parameter to allow the function's parameter serve as input and output
+	to/from the function. (Input only parameters "called by reference" not "called
+	by value" are usually declared as such by applying `const` to the parameter
+	`void func(const int& inputParameter)`. Then the remaining uncertainty is
```
Comment (wilzbach) [inline cpptod.dd]: It's `func` here, would be a valid name if you are consistent ;-)

File: cpptod.dd [hunk 20]
```diff
@@ -829,6 +831,106 @@ void test()
...
+	wether the parameter is an output only or an input&output paramater.)
```
Comment (wilzbach) [inline cpptod.dd]: `s/wether/whether`

Comment (wilzbach) [inline cpptod.dd]: "input&output" looks weird here - "input and output" isn't much better, but a bit ...

File: cpptod.dd (hunk 1 above)
Comment (fkromer) [inline cpptod.dd]: Hi @wilzbach, D is a incredibly powerful but easy to use language. It would be a shame if it would not be "promoted" accordingly... the *Programming in D for C/C++ Programmers* sections of the dlang.org were my starting point. It is fun to learn a new language and to get to know your current languages from a different perspective.

To be honest... I copy&pasted it from the above sections in the file to keep it consistent w.r.t. styling.

File: cpptod.dd (hunk 5 above)
Comment (fkromer) [inline cpptod.dd]: Oh right, thank's. (The dirty remains way longer than the quick...)

File: cpptod.dd (hunk 6 above)
Comment (fkromer) [inline cpptod.dd]: Changed for consistency

File: cpptod.dd (hunk 16 above)
Comment (fkromer) [inline cpptod.dd]: Italic looks good...

File: cpptod.dd (hunk 3 above)
Comment (fkromer) [inline cpptod.dd]: Ah, that's great I will use another time for sure!

File: cpptod.dd [hunk 21]
```diff
@@ -829,6 +831,106 @@ void test()
...
+void foo(int& inputAndOutputParameter) {
+  	inputAndOutputParameter = 42;
+}
+)
+
+The function is used as follows ($(I call-by-reference)):
+
+$(CPPCODE
+int outputArgument = 0;
+foo(int& outputArgument);
```
Comment (JinShil) [inline cpptod.dd]: shouldn't that be `foo(&outputArgument)` without the `int&`?

File: cpptod.dd [hunk 22]
```diff
@@ -829,6 +831,106 @@ void test()
...
+	$(B C++ does not support explicit input parameters.) If you use a
+	$(I pass-by-value) parameter $(D void function(int inputOnlyImplicitly)) it can just
+	be an input parameter implicitly. If you use a $(I pass-by-reference) parameter
+	usually $(D const) is applied to the parameter
+	$(D void foo(const int& inputOnlyParameterImplicitly)) that a forbidden
+	modification of the parameter in the function leads to a compilation error.
+
+$(CPPCODE
+void foo(int implicitInputParameterPassedByValue) {
+		// some modification dependent on implicitInputParameterPassedByValue
```
Comment (JinShil) [inline cpptod.dd]: Doesn't always need to be a "modification".  I suggest "some code dependent on implicitInputParameterPassedByValue".

File: cpptod.dd [hunk 23]
```diff
@@ -829,6 +831,106 @@ void test()
...
+		...
+}
+
+void foo(int& implicitInputParameterPassedByReference) {
+		// some modification dependent on implicitInputParameterPassedByReference
```
Comment (JinShil) [inline cpptod.dd]: Again, doesn't need to be a "modification".

File: cpptod.dd [hunk 24]
```diff
@@ -829,6 +831,106 @@ void test()
...
+// inputArgumentPassedByReference = 42
+)
+
+$(H4 The D Way)
+
+	D supports explicit input parameters.
+
+------
+void foo(in int inputParameter) {
+		// some modification dependent on inputParameter
```
Comment (JinShil) [inline cpptod.dd]: Again, doesn't need to be a "modification"

File: cpptod.dd [hunk 25]
```diff
@@ -829,6 +831,106 @@ void test()
...
+------
+
+The function is used as follows:
+
+------
+int inputArgument = 42;
+foo(inputArgument);
+assert(inputArgument == 42);
+------
+
```
Comment (JinShil) [inline cpptod.dd]: Could use an example showing what might happen if an attempt to modify inputArgument in foo, is attempted.

File: cpptod.dd (hunk 21 above)
Comment (fkromer) [inline cpptod.dd]: Ouuuh, right.

File: cpptod.dd (hunk 22 above)
Comment (fkromer) [inline cpptod.dd]: That's right.

File: cpptod.dd (hunk 23 above)
Comment (fkromer) [inline cpptod.dd]: Ditto.

File: cpptod.dd (hunk 24 above)
Comment (fkromer) [inline cpptod.dd]: Ditto.

File: cpptod.dd (hunk 25 above)
Comment (fkromer) [inline cpptod.dd]: Still todo.

File: cpptod.dd (hunk 3 above)
Comment (fkromer) [inline cpptod.dd]: @wilzbach I added the runnable example. I hope it works...

File: cpptod.dd (hunk 25 above)
Comment (fkromer) [inline cpptod.dd]: @JinShil I added the runnable examples. I hope it works...

File: cpptod.dd [hunk 26]
```diff
@@ -829,6 +831,124 @@ void test()
...
+)
+
+<hr>$(COMMENT  -------------------------------------------- )
+
+$(LNAME2 functionoutputparameters, Function Output Parameters)
+
+$(H4 The C++ Way)
+
+	$(B C++ does not support explicit output parameters.) You can use a reference
+	parameter to allow the function's parameter serve as input and output
```
Comment (JinShil) [inline cpptod.dd]: ... parameter serve ...  --> ... parameter to serve ...

File: cpptod.dd [hunk 27]
```diff
@@ -829,6 +831,124 @@ void test()
...
+	to/from the function. Input only parameters $(I called-by-reference) not
+	$(I called-by-value) are usually declared as such by applying $(D const) to the
+	parameter $(D void foo(const int& inputParameter)). Then the remaining uncertainty
+	is wether the parameter is an output only or an input and output parameter.
+
+$(CPPCODE
+void foo(int& inputAndOutputParameter) {
```
Comment (JinShil) [inline cpptod.dd]: Typically, starting curly braces go on the next line in D's documentation and source code.

File: cpptod.dd [hunk 28]
```diff
@@ -829,6 +831,124 @@ void test()
...
+foo(&outputArgument);
+// outputArgument = 42
+)
+
+$(H4 The D Way)
+
+	D supports explicit output parameters.
+
+------
+void foo(out int outputParameter) {
```
Comment (JinShil) [inline cpptod.dd]: Typically, starting curly braces go on the next line in D's documentation and source code.

File: cpptod.dd [hunk 29]
```diff
@@ -829,6 +831,124 @@ void test()
...
+    outputParameter = 42;
+}
+------
+
+The function is used as follows:
+
+$(RUNNABLE_EXAMPLE
+---------
+void foo(out int outputParameter) {
```
Comment (JinShil) [inline cpptod.dd]: Typically, starting curly braces go on the next line in D's documentation and source code.

File: cpptod.dd [hunk 30]
```diff
@@ -829,6 +831,124 @@ void test()
...
+		// some code dependent on inputParameter
+		...
+}
+------
+
+The function is used as follows:
+
+$(RUNNABLE_EXAMPLE
+---------
+void foo(in int inputParameter) {
```
Comment (JinShil) [inline cpptod.dd]: Typically, starting curly braces go on the next line in D's documentation and source code.

File: cpptod.dd [hunk 31]
```diff
@@ -829,6 +831,124 @@ void test()
...
+foo(const int& inputArgumentPassedByReference);
+// inputArgumentPassedByReference = 42
+)
+
+$(H4 The D Way)
+
+	D supports explicit input parameters.
+
+------
+void foo(in int inputParameter) {
```
Comment (JinShil) [inline cpptod.dd]: Typically, starting curly braces go on the next line in D's documentation and source code.

File: cpptod.dd [hunk 32]
```diff
@@ -829,6 +831,124 @@ void test()
...
+	$(D void foo(const int& inputOnlyParameterImplicitly)) that a forbidden
+	modification of the parameter in the function leads to a compilation error.
+
+$(CPPCODE
+void foo(int implicitInputParameterPassedByValue) {
+		// some code dependent on implicitInputParameterPassedByValue
+		...
+}
+
+void foo(int& implicitInputParameterPassedByReference) {
```
Comment (JinShil) [inline cpptod.dd]: Typically, starting curly braces go on the next line in D's documentation and source code.

File: cpptod.dd [hunk 33]
```diff
@@ -829,6 +831,124 @@ void test()
...
+
+	$(B C++ does not support explicit input parameters.) If you use a
+	$(I pass-by-value) parameter $(D void function(int inputOnlyImplicitly)) it can just
+	be an input parameter implicitly. If you use a $(I pass-by-reference) parameter
+	usually $(D const) is applied to the parameter
+	$(D void foo(const int& inputOnlyParameterImplicitly)) that a forbidden
+	modification of the parameter in the function leads to a compilation error.
+
+$(CPPCODE
+void foo(int implicitInputParameterPassedByValue) {
```
Comment (JinShil) [inline cpptod.dd]: Typically, starting curly braces go on the next line in D's documentation and source code.

File: cpptod.dd (hunk 26 above)
Comment (fkromer) [inline cpptod.dd]: done

File: cpptod.dd (hunk 27 above)
Comment (fkromer) [inline cpptod.dd]: done

File: cpptod.dd (hunk 33 above)
Comment (fkromer) [inline cpptod.dd]: ditto

File: cpptod.dd (hunk 32 above)
Comment (fkromer) [inline cpptod.dd]: ditto

File: cpptod.dd (hunk 31 above)
Comment (fkromer) [inline cpptod.dd]: ditto

File: cpptod.dd (hunk 30 above)
Comment (fkromer) [inline cpptod.dd]: ditto

File: cpptod.dd (hunk 28 above)
Comment (fkromer) [inline cpptod.dd]: ditto

File: cpptod.dd (hunk 29 above)
Comment (fkromer) [inline cpptod.dd]: ditto

**Comment by fkromer** [comment]
@wilzbach @JinShil Should be ok now. (I fixed the invocation of the C++ functions. They are defined as call-by-reference but where called as called-by-pointer first...)

**Comment by fkromer** [comment]
@CyberShadow conflict resolved, header syntax improved

File: cpptod.dd [hunk 34]
```diff
@@ -912,6 +914,131 @@ FirstVariant.method();
...
+
+	$(B C++ does not support explicit input parameters.) If you use a
+	$(I pass-by-value) parameter $(D void function(int inputOnlyImplicitly)) it can just
+	be an input parameter implicitly. If you use a $(I pass-by-reference) parameter
+	usually $(D const) is applied to the parameter
+	$(D void foo(const int& inputOnlyParameterImplicitly)) that a forbidden
+	modification of the parameter in the function leads to a compilation error.
+
+$(CPPCODE
+void foo(int implicitInputParameterPassedByValue)
```
Comment (wilzbach) [inline cpptod.dd]: How about `inputByValue,` `inputParamByValue`, `inputParamVal`, or `inputParameterVal`?

File: cpptod.dd [hunk 35]
```diff
@@ -912,6 +914,131 @@ FirstVariant.method();
...
+{
+		// some code dependent on implicitInputParameterPassedByValue
+		...
```
Comment (wilzbach) [inline cpptod.dd]: No tabs. Check the output. Tabs lead to unexpected/different results.

File: cpptod.dd [hunk 36]
```diff
@@ -912,6 +914,131 @@ FirstVariant.method();
...
+}
+------
+
+The function is used as follows:
+
+$(RUNNABLE_EXAMPLE
+---------
+void foo(in int inputParameter)
+{
+    ++inputParameter;
```
Comment (wilzbach) [inline cpptod.dd]: Do you want to point this out?

```
onlineapp.d(3): Error: cannot modify const expression inputParameter
```

The better mark it in the code, s.t. a reader doesn't get confused?

File: cpptod.dd [hunk 37]
```diff
@@ -25,7 +25,9 @@ $(UL
 	$(LI $(RELATIVE_LINK2 recursivetemplates, Recursive Templates))
 	$(LI $(RELATIVE_LINK2 metatemplates, Meta Templates))
 	$(LI $(RELATIVE_LINK2 typetraits, Type Traits))
-	$(LI $(RELATIVE_LINK2 interfaces, Interfaces))
+  $(LI $(RELATIVE_LINK2 interfaces, Interfaces))
```
Comment (wilzbach) [inline cpptod.dd]: Why did you change the indentation?

**Comment by carun** [comment]
@wilzbach I can help with this. I see that the author has deleted his original repo. Is there a way to pull this patch-set to work on?

**Comment by carun** [comment]
Figured out a way.. https://raw.githubusercontent.com/dlang/dlang.org/db01fae4043697d37058fc674e6cd0bcefdcf512/cpptod.dd

**Comment by fkromer** [comment]
@carun The branch didn't survive my last github repo cleanup. Luckily you were able to restore the file. Feel free to continue this contribution.

**Comment by carun** [comment]
@fkromer Pretty common with git, I hear. :-) I'm from the mercurial `hg` planet, new to git and github, so I won't be surprised if it happens to me.

**Comment by wilzbach** [comment]
Superseded by https://github.com/dlang/dlang.org/pull/2075
